lib/withOnyx.d.ts

```diff
@@ -0,0 +1,36 @@
+import {IsEqual} from 'type-fest';
+import {GetOnyxValue, OnyxKey, Selector} from './types';
+
+type BaseMapping<TComponentProps, TOnyxProps> = {
+    canEvict?: boolean | ((props: Omit<TComponentProps, keyof TOnyxProps>) => boolean);
+    initWithStoredValues?: boolean;
```

Comment:
Shouldn't we also include the following properties since they are always accessed in some way?
`key, connectionID`

Reply:
@neil-marcellini The `key` property is being defined in the `Mapping` type below. I didn't add `connectionID` because from what I understood it's being passed internally inside `withOnyx` code, so it wouldn't make sense to allow using `connectionID` in the HOC. But if we have cases where `connectionID` can be used in `withOnyx` HOC I can add as well.

Reply:
I think we always pass the `key` so it seems like it should be part of a "BaseMapping".



> I didn't add `connectionID` because from what I understood it's being passed internally inside `withOnyx` code, so it wouldn't make sense to expose `connectionID` to the developer.

Oh yeah, makes sense. I keep forgetting that we're only aiming to provide type declarations for things used externally.

Reply:
@neil-marcellini the reason that key is not part of `BaseMapping` is that there are some vacations for `key`.

It might be a function or it might be a string.

These variations are handled by `BaseMappingKey`, `BaseMappingStringKeyAndSelector` and `BaseMappingFunctionKeyAndSelector`

Reply:
wait...

@fabioh8010 what is the reason that we have union of three key types here?

```
        | BaseMappingKey<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey, OnyxEntry<KeyValueMapping[TOnyxKey]>>
        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey>
        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey>
```

what is the reason that we can't just use

```
? {
          key: TOnyxKey | ((props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey);
          selector?: Selector<TOnyxKey, TOnyxProps[TOnyxProp]>;
      }
    : never;
```